jax/_src/cudnn/fused_attention_stablehlo.py

```diff
@@ -178,42 +182,50 @@ def create_dot_product_attention_backend_config(batch,
 def get_custom_call_name(has_bias, has_mask, has_dropout, is_bwd):
   return _custom_name_maps[(is_bwd, has_dropout, has_mask, has_bias)]
 
-def check_qkv_layout(query, key, value):
-  assert len(query.shape) == len(key.shape) == len(value.shape) == 4, \
-    "query, key and value should have rank 4."
-
-  # Only support fp16 and bf16 here
-  query_dtype = query.dtype
-  key_dtype = key.dtype
-  value_dtype = value.dtype
-  assert query_dtype == key_dtype == value_dtype and query_dtype in [jnp.float16, jnp.bfloat16], \
-    "query, key and value should have same dtype and should be float16 or bfloat16"
-
-  q_batch, q_seq_len, q_num_heads, q_head_dim = query.shape
-  k_batch, k_seq_len, k_num_heads, k_head_dim = key.shape
-  v_batch, v_seq_len, v_num_heads, v_head_dim = value.shape
-  if not((q_batch == k_batch == v_batch)
-      and (k_seq_len == v_seq_len)
-      and (q_num_heads == k_num_heads == v_num_heads)
-      and (q_head_dim == k_head_dim == v_head_dim)):
-    raise ValueError(
-      "query should have layout [batch, q_seq, num_heads, head_dim], " \
-      "key and value should have layout [batch, kv_seq, num_heads, head_dim].")
-
-def check_is_flash_attention(query, key):
-  batch, q_seq_len, num_heads, head_dim = query.shape
-  _, kv_sqe_len, _, _ = key.shape
-  is_cross_attention = q_seq_len != kv_sqe_len
-  # check if attention pattern is supported by flash attention or fused attention
-  if q_seq_len > 512 and kv_sqe_len > 512 and head_dim in [64, 128]:
-    # check if flash attention is supported
+def check_qkv_layout(query, key, value, layout):
+  def assert_eq(a, b, c, msg):
+    assert a == b == c, msg + f' must be same: {a}, {b}, {c}'
+
+  Q_rank, K_rank, V_rank  = len(query.shape), len(key.shape), len(value.shape)
+  assert Q_rank == 4, "Q must have a rank of 4"
+  assert_eq(Q_rank, K_rank, V_rank, 'QKV rank')
+
+  Q_dtype, K_dtype, V_dtype = query.dtype, key.dtype, value.dtype
+  assert Q_dtype in [jnp.float16, jnp.bfloat16], "Q must be fp16 or bf16"
+  assert_eq(Q_dtype, K_dtype, V_dtype, 'QKV dtype')
+
+  if layout == 'BNTH':
+    qB, qN,  _, qH = query.shape
+    kB, kN, kS, kH = key.shape
+    vB, vN, vS, vH = value.shape
+  else:
+    assert layout == 'BTNH'
+    qB,  _, qN, qH = query.shape
+    kB, kS, kN, kH = key.shape
+    vB, vS, vN, vH = value.shape
+  
+  assert_eq(qB, kB, vB, 'QKV batch')
```

Comment:
For this part, i think the original implementation is good enough, let's keep unrelated changes minimum.

Reply:
To support a different layout, we need to manipulate the names of different dims. So, I make them more concise and remove duplicate code. So, I think this is related.